File: frontend/src/libs/Rating.ts
import { ColorCode } from "./ColorCode";
import type { Optional } from "./types";

type RawRating =
  | "bad"
  | "below_avg"
  | "avg"
  | "good"
  | "very_good"
  | "great"
  | "unicum"
  | "super_unicum";

type RatingThreshold = {
  readonly raw: RawRating;
  readonly pr: number;
  readonly shipDamageRatio: number;
  readonly winRate: number;
};

const THRESHOLDS: RatingThreshold[] = [
  {
    raw: "bad",
    pr: 0,
    shipDamageRatio: 0,
    winRate: 0,
  },
  {
    raw: "below_avg",
    pr: 750,
    shipDamageRatio: 0.6,
    winRate: 47,
  },
  {
    raw: "avg",
    pr: 1100,
    shipDamageRatio: 0.8,
    winRate: 50,
  },
  {
    raw: "good",
    pr: 1350,
    shipDamageRatio: 1.0,
    winRate: 52,
  },
  {
    raw: "very_good",
    pr: 1550,
    shipDamageRatio: 1.2,
    winRate: 54,
  },
  {
    raw: "great",
    pr: 1750,
    shipDamageRatio: 1.4,
    winRate: 56,
  },
  {
    raw: "unicum",
    pr: 2100,
    shipDamageRatio: 1.5,
    winRate: 60,
  },
  {
    raw: "super_unicum",
    pr: 2450,
    shipDamageRatio: 1.6,
    winRate: 65,
  },
] as const;

const COLORS: { [raw in RawRating]: ColorCode } = {
  bad: new ColorCode("#FE0E00"),
  below_avg: new ColorCode("#FE7903"),
  avg: new ColorCode("#FFC71F"),
  good: new ColorCode("#44B300"),
  very_good: new ColorCode("#318000"),
  great: new ColorCode("#02C9B3"),
  unicum: new ColorCode("#D042F3"),
  super_unicum: new ColorCode("#A00DC5"),
} as const;

const DISPLAY_NAMES: { [raw in RawRating]: string } = {
  bad: "Bad",
  below_avg: "Below Average",
  avg: "Average",
  good: "Good",
  very_good: "Very Good",
  great: "Great",
  unicum: "Unicum",
  super_unicum: "Super Unicum",
} as const;

export class Rating {
  constructor(public readonly raw: RawRating) {}

  static fromPR(value: number): Optional<Rating> {
    const satisfied = THRESHOLDS.findLast((it) => value >= it.pr);
    if (!satisfied) {
      return undefined;
    }
    return new Rating(satisfied.raw);
  }

  static fromWinRate(value: number): Optional<Rating> {
    const satisfied = THRESHOLDS.findLast((it) => value >= it.winRate);
    if (!satisfied) {
      return undefined;
    }
    return new Rating(satisfied.raw);
  }

  static fromShipDamage = (
    value: number,
    expected: number,
  ): Optional<Rating> => {
    if (expected <= 0) {
      return undefined;
    }
    const ratio = value / expected;

    const satisfied = THRESHOLDS.findLast((it) => ratio >= it.shipDamageRatio);
    if (!satisfied) {
      return undefined;
    }
    return new Rating(satisfied.raw);
  };

  static getThresholds(): RatingThreshold[] {
    return THRESHOLDS;
  }

  getColor(): ColorCode {
    return COLORS[this.raw];
  }

  getDisplayName(): string {
    return DISPLAY_NAMES[this.raw];
  }
}

File: backend/data/rating.go
package data

type Rating string

const (
	RatingSuperUnicum Rating = "super_unicum"
	RatingUnicum      Rating = "unicum"
	RatingGreat       Rating = "great"
	RatingVeryGood    Rating = "very_good"
	RatingGood        Rating = "good"
	RatingAvg         Rating = "avg"
	RatingBelowAvg    Rating = "below_avg"
	RatingBad         Rating = "bad"
	RatingNone        Rating = "none"
)

func NewDamageRatings(expect float64) []RatingValue {
	var result = make([]RatingValue, 0, len(thresholds))
	for _, v := range thresholds {
		value := expect * v.ShipDamageRatio
		result = append(result, RatingValue{
			Value:  value,
			Rating: v.Rating,
		})
	}

	return result
}

type RatingThreshold struct {
	Rating          Rating
	PR              float64
	ShipDamageRatio float64
	WinRate         float64
}

//nolint:gochecknoglobals
var thresholds = []RatingThreshold{
	{Rating: RatingSuperUnicum, PR: 2450, ShipDamageRatio: 1.6, WinRate: 65},
	{Rating: RatingUnicum, PR: 2100, ShipDamageRatio: 1.5, WinRate: 60},
	{Rating: RatingGreat, PR: 1750, ShipDamageRatio: 1.4, WinRate: 56},
	{Rating: RatingVeryGood, PR: 1550, ShipDamageRatio: 1.2, WinRate: 54},
	{Rating: RatingGood, PR: 1350, ShipDamageRatio: 1.0, WinRate: 52},
	{Rating: RatingAvg, PR: 1100, ShipDamageRatio: 0.8, WinRate: 50},
	{Rating: RatingBelowAvg, PR: 750, ShipDamageRatio: 0.6, WinRate: 47},
	{Rating: RatingBad, PR: 0, ShipDamageRatio: 0, WinRate: 0},
}

func NewRatingFromPR(value float64) Rating {
	for _, t := range thresholds {
		if value >= t.PR {
			return t.Rating
		}
	}

	return RatingNone
}

func NewRatingFromWinRate(value float64) Rating {
	for _, t := range thresholds {
		if value >= t.WinRate {
			return t.Rating
		}
	}

	return RatingNone
}

func NewRatingFromShipDamage(value, expected float64) Rating {
	if expected <= 0 {
		return RatingNone
	}

	ratio := value / expected
	for _, t := range thresholds {
		if ratio >= t.ShipDamageRatio {
			return t.Rating
		}
	}

	return RatingNone
}
